packages/@lwc/engine-dom/src/renderer.ts

```diff
@@ -200,10 +190,8 @@ export function nextSibling(node: Node): Node | null {
 }
 
 export function attachShadow(element: Element, options: ShadowRootInit): ShadowRoot {
-    if (hydrating) {
-        return element.shadowRoot!;
-    }
-    return element.attachShadow(options);
+    const internals = (element as any).attachInternals?.();
```

Comment:
This is the most important part of the PR. We rely on attachInternals to access any shadowRoot (open or closed), and that will take care of returning the right shadow. This method in general should probably be renamed in the future to something more like: `attachShadowIfNeeded`

Reply:
`attachInternals` is [not supported in Safari](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/attachInternals). Would we need to polyfill this?

Reply:
no, not in this case. In the previous incarnation of this code, the option was always to assume that an open shadow was created, and therefor the `.shadowRoot` value was there. This one assumes that attachInternals might not be there, and then attach. An alternative here is to do the following in the next line:

```
return internals?.shadowRoot || element.shadowRoot || element.attachShadow(options);
```

which covers safari.

Reply:
So if we assume we're running in Safari, then this code would boil down to:

```js
return element.shadowRoot || element.attachShadow(options)
```

Can we use this instead of `attachInternals`? What benefit do we get from using `attachInternals` in Firefox and Chrome?

Reply:
exactly, the same old code.

now, in terms of benefits, two main benefits:

1. avoid users to poison the engine, e.g.: `class Foo extends HTMLElement { attachShadow() { return something; } }`.
2. closed shadows hydration.

Reply:
For # 1, couldn't users poison the engine using `class Foo extends HTMLElement { attachInternals() { return something; } }` instead?

For # 2, I'm a bit surprised because it seems `attachInternals` defeats the whole purpose of closed shadow DOM. But I confirmed that you're right, you can access the shadow root that way:

![Screen Shot 2022-04-28 at 3 50 22 PM](https://user-images.githubusercontent.com/283842/165860404-72aad8f2-0081-4be4-bac5-979520c8683f.png)

Reply:
Well, for #1, LWC is NOT protecting against poisoning. LWS does everywhere. You don't protect about them providing `attachShadow()` jejejeje. You can simply cache the method, and call it to avoid dot notation, but again, that's not a thing... it was a some point, but for some reason, that was removed everywhere.

For #2, NO, `attachInternals` can only be called during the construction phase... after that it throws.

Reply:
Based on my testing (see screenshot above), `attachInternals` can be called after the constructor phase. But it can't be called twice. So I guess a closed-shadow-root component would want to call it during the constructor phase so that someone else can't get access to its shadow root.

If `attachInternals` is not supported in Safari, and if there's no advantage in Chrome/Firefox to using it, then I kinda feel we are better off avoiding it… otherwise we have the added complexity of 2 code paths for 2 different browsers, with no real advantage.

Reply:
All the [hydration tests](https://app.circleci.com/pipelines/github/salesforce/lwc/5937/workflows/f2a6f16f-9014-484e-a9b2-0bb04b95a648/jobs/43359) currently fails, because the hydrated elements are never registered as custom elements.